File: S2P2_herencia/P2A4Empleado/P2A4C#/TrabajadorPiezas.cs
// Nueva clase: TrabajadorPiezas
// Representa a un empleado que cobra por pieza producida

public class TrabajadorPiezas : Empleado {
    private double salarioPorPieza; // pago por cada pieza
    private int piezas; // número de piezas producidas

    // constructor
    public TrabajadorPiezas(string nombre, string apellido, string nss,
            double salarioPorPieza, int piezas)
        : base(nombre, apellido, nss)
    {
        establecerSalarioPorPieza(salarioPorPieza);
        establecerPiezas(piezas);
    }

    // establecer salario por pieza
    public void establecerSalarioPorPieza(double salario) {
        salarioPorPieza = (salario < 0.0) ? 0.0 : salario;
    }

    // obtener salario por pieza
    public double obtenerSalarioPorPieza() {
        return salarioPorPieza;
    }

    // establecer piezas producidas
    public void establecerPiezas(int piezasProducidas) {
        piezas = (piezasProducidas < 0) ? 0 : piezasProducidas;
    }

    // obtener piezas producidas
    public int obtenerPiezas() {
        return piezas;
    }

    // implementar método abstracto ingresos
    public override double Ingresos() {
        return salarioPorPieza * piezas;
    }

    // toString
    public override string ToString() {
        return string.Format(
                "trabajador por piezas: {0}\n{1}: {2:C2}; {3}: {4}",
                base.ToString(),
                "salario por pieza", obtenerSalarioPorPieza(),
                "piezas producidas", obtenerPiezas());
    }
}

File: S2P2_herencia/P2A4Empleado/P2A4C++/TrabajadorPiezas.cpp
#include "TrabajadorPiezas.h"
#include <sstream>
#include <iomanip>

TrabajadorPiezas::TrabajadorPiezas(const std::string& nombre, const std::string& apellido, 
                                   const std::string& nss, double salarioPorPieza, int piezas)
    : Empleado(nombre, apellido, nss) {
    establecerSalarioPorPieza(salarioPorPieza);
    establecerPiezas(piezas);
}

void TrabajadorPiezas::establecerSalarioPorPieza(double salario) {
    salarioPorPieza = (salario < 0.0) ? 0.0 : salario;
}

double TrabajadorPiezas::obtenerSalarioPorPieza() const {
    return salarioPorPieza;
}

void TrabajadorPiezas::establecerPiezas(int piezasProducidas) {
    piezas = (piezasProducidas < 0) ? 0 : piezasProducidas;
}

int TrabajadorPiezas::obtenerPiezas() const {
    return piezas;
}

double TrabajadorPiezas::ingresos() const {
    return obtenerSalarioPorPieza() * obtenerPiezas();
}

std::string TrabajadorPiezas::toString() const {
    std::ostringstream output;
    output << std::fixed << std::setprecision(2);
    output << "trabajador por piezas: " << Empleado::toString()
           << "\nsalario por pieza: $" << obtenerSalarioPorPieza()
           << "; piezas producidas: " << obtenerPiezas();
    return output.str();
}
